File: 55 - Jump Game.cpp
/*
 55. Jump Game (Medium)

 Notes:
    Greedy algorithm - O(n)
    Scan the list from end to begin. If the list[i] value is "0", the list[i - 1] will need to have addtional step
    to skip this element. If reach the first element and still need steps to pay off additional steps, then it is 
    impossible to reach the end.
*/

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int gap = 0;
        for (int i = nums.size() - 2; i >= 0; i--)
        {
            if (nums[i] - gap <= 0)
                gap++;
            else
                gap = 0;
        }
        return gap == 0;
    }
};

File: JAVA/55 - Jump Game.java
/*
 55. Jump Game (M)

 Approach 1 : Backtracking   
 Scan the list from end to head. If the current scanned element is 0, it means
 the next scanned element will need to have one additional step to pass this 
 element, or it will stuck in here. If scan is over but still need steps to pay
 off, then it is impossible to reach the end.
 Time: O(n)
 Space: O(1)

 Approach 2 : DP
 Time: O(n)
 Space: O(nums.length)

*/

/* Approach 1 */
class Solution {
    public boolean canJump(int[] nums) {
        int skip = 0;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] - skip <= 0) skip++;
            else skip = 0;
        }
        return skip == 0;
    }
}

/* Approach 2 */
class Solution {
    public boolean canJump(int[] nums) {
        int[] dp = new int[nums.length];
        for (int i = 1; i < nums.length; i++) {
            dp[i] = Math.max(dp[i - 1], nums[i - 1]) - 1;
            if (dp[i] < 0) return false;
        }
        return true;        
    }
}